contracts/ETHRegistrarController.sol

```diff
@@ -123,20 +139,38 @@ contract ETHRegistrarController is Ownable {
         if(msg.value > cost) {
             msg.sender.transfer(msg.value - cost);
         }
+
+        _sendReferralFee(referrer, cost);
     }
 
     function renew(string calldata name, uint duration) external payable {
-        uint cost = rentPrice(name, duration);
-        require(msg.value >= cost);
+        renewWithReferrer(name, duration, address(0));
+    }
 
-        bytes32 label = keccak256(bytes(name));
-        uint expires = base.renew(uint256(label), duration);
+    function renewWithReferrer(string memory name, uint duration, address payable referrer) public payable {
+        uint cost = _doRenew(name, duration);
+        require(msg.value >= cost);
 
+        // Refund any extra
         if(msg.value > cost) {
             msg.sender.transfer(msg.value - cost);
         }
 
-        emit NameRenewed(name, label, cost, expires);
+        _sendReferralFee(referrer, cost);
```

Comment:
I may not be understanding the logic as you have a test to cover, but how come you still have a referrer fee when you already sent back any extra at line 156?

Reply:
The extra is the difference between the cost and what was sent. The referral fee comes out of the cost.

Reply:
Ah I see